File: 1578.java
class Solution {
    public int minCost(String colors, int[] neededTime) {
        int n = colors.length();
        int result = 0;
        int prevCost = neededTime[0];

        for (int i = 1; i < n; i++) {
            if (colors.charAt(i) == colors.charAt(i - 1)) {
                result += Math.min(neededTime[i], prevCost);
                prevCost = Math.max(neededTime[i], prevCost);
            } else
                prevCost = neededTime[i];
        }

        return result;
    }
}

File: Python/1578.py
from typing import List


class Solution:
    def minCost(self, colors: str, neededTime: List[int]) -> int:
        result = cur_max = 0
        for i in range(len(colors)):
            if i > 0 and colors[i] != colors[i - 1]:
                cur_max = 0
            result += min(cur_max, neededTime[i])
            cur_max = max(cur_max, neededTime[i])
        return result
